File: SignalServiceKit/src/TestUtils/TestAppContext.m
//
// Copyright 2018 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

#import "TestAppContext.h"
#import "OWSFileSystem.h"
#import <SignalServiceKit/SignalServiceKit-Swift.h>

NS_ASSUME_NONNULL_BEGIN

#ifdef TESTABLE_BUILD

@interface TestAppContext ()

@property (nonatomic) SSKTestKeychainStorage *testKeychainStorage;
@property (nonatomic) NSString *mockAppDocumentDirectoryPath;
@property (nonatomic) NSString *mockAppSharedDataDirectoryPath;
@property (nonatomic) NSUserDefaults *appUserDefaults;

@end

#pragma mark -

@implementation TestAppContext

@synthesize mainWindow = _mainWindow;
@synthesize appLaunchTime = _appLaunchTime;
@synthesize appForegroundTime = _appForegroundTime;

- (instancetype)init
{
    self = [super init];
    if (!self) {
        return self;
    }

    self.testKeychainStorage = [SSKTestKeychainStorage new];

    // Avoid using OWSTemporaryDirectory(); it can consult the current app context.
    NSString *dirName = [NSString stringWithFormat:@"ows_temp_%@", NSUUID.UUID.UUIDString];
    NSString *temporaryDirectory = [NSTemporaryDirectory() stringByAppendingPathComponent:dirName];
    NSError *error = nil;
    [[NSFileManager defaultManager] createDirectoryAtPath:temporaryDirectory
                              withIntermediateDirectories:YES
                                               attributes:nil
                                                    error:&error];
    if (error) {
        OWSFail(@"Failed to create directory: %@, error: %@", temporaryDirectory, error);
    }

    self.mockAppDocumentDirectoryPath = [temporaryDirectory stringByAppendingPathComponent:NSUUID.UUID.UUIDString];
    self.mockAppSharedDataDirectoryPath = [temporaryDirectory stringByAppendingPathComponent:NSUUID.UUID.UUIDString];
    self.appUserDefaults = [[NSUserDefaults alloc] init];
    NSDate *launchDate = [NSDate new];
    _appLaunchTime = launchDate;
    _appForegroundTime = launchDate;

    return self;
}

- (UIApplicationState)reportedApplicationState
{
    return UIApplicationStateActive;
}

#pragma mark -

- (BOOL)isMainApp
{
    return YES;
}

- (BOOL)isMainAppAndActive
{
    return YES;
}

- (BOOL)isNSE
{
    return NO;
}

- (UIApplicationState)mainApplicationStateOnLaunch
{
    OWSFailDebug(@"Not main app.");

    return UIApplicationStateInactive;
}

- (BOOL)isRTL
{
    return NO;
}

- (CGFloat)statusBarHeight
{
    return 20;
}

- (BOOL)isInBackground
{
    return NO;
}

- (BOOL)isAppForegroundAndActive
{
    return YES;
}

- (UIBackgroundTaskIdentifier)beginBackgroundTaskWithExpirationHandler:
    (BackgroundTaskExpirationHandler)expirationHandler
{
    return UIBackgroundTaskInvalid;
}

- (void)endBackgroundTask:(UIBackgroundTaskIdentifier)backgroundTaskIdentifier
{
}

- (void)ensureSleepBlocking:(BOOL)shouldBeBlocking blockingObjectsDescription:(NSString *)blockingObjectsDescription
{
}

- (nullable UIViewController *)frontmostViewController
{
    return nil;
}

- (void)openSystemSettings
{
}

- (void)openURL:(NSURL *)url completion:(void (^__nullable)(BOOL))completion
{
}

- (BOOL)isRunningTests
{
    return YES;
}

- (CGRect)frame
{
    // Pretend to be a small device.
    return CGRectMake(0, 0, 300, 400);
}

- (UIInterfaceOrientation)interfaceOrientation
{
    return UIInterfaceOrientationPortrait;
}

#pragma mark -

- (void)runNowOrWhenMainAppIsActive:(AppActiveBlock)block
{
    block();
}

- (void)runAppActiveBlocks
{
}

- (id<SSKKeychainStorage>)keychainStorage
{
    return self.testKeychainStorage;
}

- (NSString *)appDocumentDirectoryPath
{
    return self.mockAppDocumentDirectoryPath;
}

- (NSString *)appSharedDataDirectoryPath
{
    return self.mockAppSharedDataDirectoryPath;
}

- (NSString *)appDatabaseBaseDirectoryPath
{
    return self.appSharedDataDirectoryPath;
}

- (BOOL)canPresentNotifications
{
    return NO;
}

- (BOOL)shouldProcessIncomingMessages
{
    return YES;
}

- (BOOL)hasUI
{
    return YES;
}

- (BOOL)hasActiveCall
{
    return NO;
}

- (NSString *)debugLogsDirPath
{
    return TestAppContext.testDebugLogsDirPath;
}

+ (NSString *)testDebugLogsDirPath
{
    NSString *dirPath = [OWSTemporaryDirectory() stringByAppendingPathComponent:@"TestLogs"];
    [OWSFileSystem ensureDirectoryExists:dirPath];
    return dirPath;
}

- (void)resetAppDataAndExit
{
    // Do nothing.
}

@end

#endif

NS_ASSUME_NONNULL_END

File: SignalServiceKit/TestUtils/TestAppContext.swift
//
// Copyright 2024 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation

#if TESTABLE_BUILD

@objc
public class TestAppContext: NSObject, AppContext {
    public static var testDebugLogsDirPath: String {
        let dirPath = OWSTemporaryDirectory().appendingPathComponent("TestLogs")
        OWSFileSystem.ensureDirectoryExists(dirPath)
        return dirPath
    }

    private let mockAppDocumentDirectoryPath: String
    private let mockAppSharedDataDirectoryPath: String
    private let internalAppUserDefaults: UserDefaults
    public func appUserDefaults() -> UserDefaults { internalAppUserDefaults }

    // MARK: -

    public var mainWindow: UIWindow?
    public let appLaunchTime: Date
    public let appForegroundTime: Date

    public override init() {
        // Avoid using OWSTemporaryDirectory(); it can consult the current app context.
        let dirName = "ows_temp_\(UUID().uuidString)"
        let temporaryDirectory = NSTemporaryDirectory().appendingPathComponent(dirName)
        do {
            try FileManager.default.createDirectory(atPath: temporaryDirectory, withIntermediateDirectories: true)
        } catch {
            owsFail("Failed to create directory: \(temporaryDirectory), error: \(error)")
        }

        self.mockAppDocumentDirectoryPath = temporaryDirectory.appendingPathComponent(UUID().uuidString)
        self.mockAppSharedDataDirectoryPath = temporaryDirectory.appendingPathComponent(UUID().uuidString)
        self.internalAppUserDefaults = UserDefaults()
        let launchDate = Date()
        self.appLaunchTime = launchDate
        self.appForegroundTime = launchDate

        super.init()
    }

    public var reportedApplicationState: UIApplication.State = .active

    // MARK: -

    public let type: SignalServiceKit.AppContextType = .main
    public let isMainAppAndActive: Bool = true
    public let isMainAppAndActiveIsolated: Bool = true
    public func mainApplicationStateOnLaunch() -> UIApplication.State { .inactive }
    public let isRTL: Bool = false
    public func isInBackground() -> Bool { false }
    public func isAppForegroundAndActive() -> Bool { true }
    public func beginBackgroundTask(with expirationHandler: BackgroundTaskExpirationHandler) -> UIBackgroundTaskIdentifier { .invalid }
    public func endBackgroundTask(_ backgroundTaskIdentifier: UIBackgroundTaskIdentifier) {}
    public func ensureSleepBlocking(_ shouldBeBlocking: Bool, blockingObjectsDescription: String) {}
    public func frontmostViewController() -> UIViewController? { nil }
    public func openSystemSettings() {}
    public func open(_ url: URL, completion: ((Bool) -> Void)?) {}
    public let isRunningTests: Bool = true
    /// Pretend to be a small device.
    public let frame: CGRect = CGRect(x: 0, y: 0, width: 300, height: 400)

    // MARK: -

    public func runNowOrWhenMainAppIsActive(_ block: AppActiveBlock) { block() }
    public func appDocumentDirectoryPath() -> String { mockAppDocumentDirectoryPath }
    public func appSharedDataDirectoryPath() -> String { mockAppSharedDataDirectoryPath }
    public func appDatabaseBaseDirectoryPath() -> String { appSharedDataDirectoryPath() }
    public func canPresentNotifications() -> Bool { false }
    public let shouldProcessIncomingMessages: Bool = true
    public let hasUI: Bool = true
    public let debugLogsDirPath: String = testDebugLogsDirPath

    @MainActor
    public func resetAppDataAndExit() -> Never {
        owsFail("resetAppDataAndExit called during tests")
    }
}

#endif
